File: plugins/de.ovgu.featureide.examples/featureide_examples/GPL-FH-JML/features/DirectedWithEdges/GPL/Graph.java
package GPL;

import java.util.LinkedList;
import java.util.Iterator;
import java.util.Collections;
import java.util.Comparator;

// *************************************************************************

public class Graph {
    private /*@spec_public@*/ LinkedList vertices;
    private /*@spec_public@*/ LinkedList edges;
    public /*@spec_public@*/ static final boolean isDirected = true;

    /*@ensures vertices != null && edges != null;@*/
    public Graph() {
        vertices = new LinkedList();
        edges = new LinkedList();
    }

    // Fall back method that stops the execution of programs
    public void run( Vertex s ) {}

    /*@ requires c != null;
    ensures isSorted(edges); @*/
    public void sortEdges(Comparator c) {
        Collections.sort(edges, c);
    }
	/*@public model pure boolean isSorted(LinkedList list) {
 	return true;
 	}@*/
    
    
    
    /*@ requires c != null;
    ensures isSorted(vertices); @*/
    public void sortVertices(Comparator c) {
        Collections.sort(vertices, c);
    }

    // Adds an edge without weights if Weighted layer is not present
	/*@ requires start != null && end != null; @*/
	/*@ ensures \result.getOtherVertex(start)==end && \result.getOtherVertex(end)==start; @*/
    public EdgeIfc addEdge(Vertex start,  Vertex end) {
        Edge theEdge = new  Edge();
        theEdge.EdgeConstructor( start, end );
        edges.add( theEdge );
        start.addNeighbor( new  Neighbor( end, theEdge ) );
        //end.addNeighbor( new  Neighbor( start, theEdge ) );

        return theEdge;
    }

    /*@requires v != null;@*/
    /*@ensures vertices.getLast()==v;@*/
    protected void addVertex( Vertex v ) {
        vertices.add( v );
    }

    // Finds a vertex given its name in the vertices list
    /*@ ensures (theName == null) ==> (\result == null); @*/
    /*@ ensures (\forall int i; i >= 0 && i < vertices.size; (vertices.get(i)!=null) ==> (vertices.get(i).getName().equals(theName) ==> \result != null)); @*/
    /*@ ensures (\result != null) ==> \result.getName() == theName; @*/
    public /*@pure@*/  Vertex findsVertex( String theName )
      {
        Vertex theVertex = null;

        // if we are dealing with the root
        if ( theName==null )
            return null;

        for(VertexIter vxiter = getVertices(); vxiter.hasNext(); )
        {
            theVertex = vxiter.next();
            if ( theName.equals( theVertex.getName() ) )
                return theVertex;
        }

        return theVertex;
    }
    /*@ ensures \result != null; @*/
    public VertexIter getVertices() {
        return new VertexIter() {
                private Iterator iter = vertices.iterator();
                public Vertex next() { return (Vertex)iter.next(); }
                public boolean hasNext() { return iter.hasNext(); }
            };
    }


    public EdgeIter getEdges() {
        return new EdgeIter() {
                private Iterator iter = edges.iterator();
                public EdgeIfc next() { return (EdgeIfc)iter.next(); }
                public boolean hasNext() { return iter.hasNext(); }
            };
    }

    public void display() {
        int i;

        System.out.println( "******************************************" );
        System.out.println( "Vertices " );
        for ( VertexIter vxiter = getVertices(); vxiter.hasNext() ; )
            vxiter.next().display();

        System.out.println( "******************************************" );
        System.out.println( "Edges " );
        for ( EdgeIter edgeiter = getEdges(); edgeiter.hasNext(); )
            edgeiter.next().display();

        System.out.println( "******************************************" );
    }

}

File: plugins/de.ovgu.featureide.examples/featureide_examples/GPL-FH-CSharp/features/DirectedWithEdges/Graph.cs
//layer Directed;

using System.Collections;
using System.Collections.Generic;
//import java.util.Iterator;
//import java.util.Collections;
//import java.util.Comparator;

// *************************************************************************

public class Graph {
    private List<Vertex> vertices;
    private List<EdgeIfc> edges;
    public static readonly bool isDirected = true;

    public Graph() {
        vertices = new List<Vertex>();
        edges = new List<EdgeIfc>();
    }

    // Fall back method that stops the execution of programs
    public void run( Vertex s ) {}

    public void sortEdges(System.Collections.Generic.IComparer<EdgeIfc> c) {
    	edges.Sort(c);
    }

    public void sortVertices(System.Collections.Generic.IComparer<Vertex> c) {
    	vertices.Sort(c);
    }

    // Adds an edge without weights if Weighted layer is not present
    public EdgeIfc AddEdge(Vertex start,  Vertex end) {
        Edge theEdge = new  Edge();
        theEdge.EdgeConstructor( start, end );
        edges.Add( theEdge );
        start.AddNeighbor( new  Neighbor( end, theEdge ) );
        //end.AddNeighbor( new  Neighbor( start, theEdge ) );

        return theEdge;
    }

    public void AddVertex( Vertex v ) {
        vertices.Add( v );
    }

    // Finds a vertex given its name in the vertices list
    public  Vertex findsVertex( string theName )
      {
        Vertex theVertex = null;

        // if we are dealing with the root
        if ( theName==null )
            return null;

        for(Iterator<Vertex> vxiter = GetVertices(); vxiter.hasNext(); )
        {
            theVertex = vxiter.next();
            if ( theName.Equals( theVertex.GetName() ) )
                return theVertex;
        }

        return theVertex;
    }

    public Iterator<Vertex> GetVertices() {
        return new Iterator<Vertex>(vertices.GetEnumerator());
    }
    public Iterator<EdgeIfc> GetEdges() {
        return new Iterator<EdgeIfc>(edges.GetEnumerator());
    }
    public void display() {
        int i;

        System.Console.Out.WriteLine( "******************************************" );
        System.Console.Out.WriteLine( "Vertices " );
        for ( Iterator<Vertex> vxiter = GetVertices(); vxiter.hasNext() ; )
            vxiter.next().display();

        System.Console.Out.WriteLine( "******************************************" );
        System.Console.Out.WriteLine( "Edges " );
        for ( Iterator<EdgeIfc> edgeiter = GetEdges(); edgeiter.hasNext(); )
            edgeiter.next().display();

        System.Console.Out.WriteLine( "******************************************" );
    }

}
